src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonDictionaryConverter.cs

```diff
@@ -230,14 +230,14 @@ protected static JsonConverter<T> GetConverter<T>(JsonTypeInfo typeInfo)
                 {
                     if (state.Current.PropertyState == StackFramePropertyState.None)
                     {
-                        state.Current.PropertyState = StackFramePropertyState.ReadName;
-
                         // Read the key name.
                         if (!reader.Read())
                         {
                             value = default;
                             return false;
                         }
+
```

Comment:
Changing the order of state update operations since we no longer rely on the root method to advance the reader for us on resumed operations.